StandaloneMmPkg/Drivers/StandaloneMmIplPei/StandaloneMmIplPei.c

```diff
@@ -0,0 +1,800 @@
+  //
+  Status = LocateMmCoreFv (&MmFvBase, &MmFvSize, &MmCoreFileName, &ImageContext.Handle);
+  ASSERT_EFI_ERROR (Status);
+
+  //
+  // Unblock the MM FV range to be accessible from inside MM
+  //
+  if ((MmFvBase != 0) && (MmFvSize != 0)) {
+    Status = MmUnblockMemoryRequest (MmFvBase, EFI_SIZE_TO_PAGES (MmFvSize));
+    ASSERT_EFI_ERROR (Status);
+  }
+
+  //
+  // Initialize ImageContext
+  //
+  ImageContext.ImageRead = PeCoffLoaderImageReadFromMemory;
+
+  //
+  // Get information about the image being loaded
+  //
+  Status = PeCoffLoaderGetImageInfo (&ImageContext);
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  PageCount = (UINTN)EFI_SIZE_TO_PAGES ((UINTN)ImageContext.ImageSize + ImageContext.SectionAlignment);
+
+  //
+  // Allocate memory for the image being loaded from unallocated mmram range
+  //
+  ImageContext.ImageAddress = MmIplAllocateMmramPage (PageCount, &Block);
+  if (ImageContext.ImageAddress == 0) {
+    return EFI_NOT_FOUND;
+  }
+
+  //
+  // Align buffer on section boundary
+  //
+  ImageContext.ImageAddress += ImageContext.SectionAlignment - 1;
+  ImageContext.ImageAddress &= ~((EFI_PHYSICAL_ADDRESS)ImageContext.SectionAlignment - 1);
+
+  //
+  // Print debug message showing MM Core load address.
+  //
+  DEBUG ((DEBUG_INFO, "StandaloneMM IPL loading MM Core at MMRAM address %p\n", (VOID *)(UINTN)ImageContext.ImageAddress));
+
+  //
+  // Load the image to our new buffer
+  //
+  Status = PeCoffLoaderLoadImage (&ImageContext);
+  if (!EFI_ERROR (Status)) {
+    //
+    // Relocate the image in our new buffer
+    //
+    Status = PeCoffLoaderRelocateImage (&ImageContext);
+    if (!EFI_ERROR (Status)) {
+      DEBUG ((DEBUG_INFO, "MmCoreImageBase  - 0x%016lx\n", ImageContext.ImageAddress));
+      DEBUG ((DEBUG_INFO, "MmCoreImageSize  - 0x%016lx\n", ImageContext.ImageSize));
+
+      //
+      // Flush the instruction cache so the image data are written before we execute it
+      //
+      InvalidateInstructionCacheRange ((VOID *)(UINTN)ImageContext.ImageAddress, (UINTN)ImageContext.ImageSize);
+
+      //
+      // Get HOB list for Standalone MM Core.
+      //
+      MmHobSize = 0;
+      MmHobList = CreatMmHobList (
+                    &MmHobSize,
+                    MmCommBuffer,
+                    MmFvBase,
+                    MmFvSize,
+                    &MmCoreFileName,
+                    ImageContext.ImageAddress,
+                    ImageContext.ImageSize,
+                    ImageContext.EntryPoint,
+                    Block
+                    );
+
+      //
+      // Print debug message showing Standalone MM Core entry point address.
+      //
+      DEBUG ((DEBUG_INFO, "StandaloneMM IPL calling Standalone MM Core at MMRAM address - 0x%016lx\n", ImageContext.EntryPoint));
+
+      //
+      // Execute image
+      //
+      Entry  = (STANDALONE_MM_FOUNDATION_ENTRY_POINT)(UINTN)ImageContext.EntryPoint;
+      Status = Entry (MmHobList);

```

Comment:
Thanks for upstreaming this design. But how would this entry work if the standalone mm core is x64 yet the PEIM here is running in IA32 bit mode? Should there be a thunking to 64 bit mode before dispatching the mm core?

Reply:
This is a X64 PEIM driver, so it is no need to a thunking to 64 bit mode.
For IA32 PEIM driver, there is no plan to support it now.

Reply:
I see. Then maybe remove the `IA32` in the inf?

Reply:
Yes